onnxruntime/core/optimizer/unsqueeze_elimination.cc

```diff
@@ -2,104 +2,108 @@
-      if (axis == 0) {
-        axis = *begin++;
-      }
+  for (int i = 0; i < static_cast<int>(new_dims.size()); i++) {
+    if (i < tensor_proto->dims().size()) {
+      new_tensor_proto.set_dims(i, new_dims[i]);
+    } else {
+      new_tensor_proto.add_dims(new_dims[i]);
     }
+  }
+  graph.RemoveInitializedTensor(input_def->Name());
+  graph.AddInitializedTensor(new_tensor_proto);
 
-    // Update shape of tensor proto
-    ONNX_NAMESPACE::TensorProto new_tensor_proto(*tensor_proto);
+  // Update shape of NodeArg
+  TensorShapeProto shape;
+  for (auto dim : new_dims) {
+    shape.add_dim()->set_dim_value(dim);
+  }
+  input_def->SetShape(shape);
+
+  // Keep all output nodes in a list (after removing all edges below, there will be no other way to access them; 
+  // I cannot first update the output defs, because edge removal will not work afterwards).
+  std::vector<NodeIndex*> output_nodes_idx;
+  for (auto it = node.OutputNodesBegin(), end = node.OutputNodesEnd(); it != end; ++it) {
+    auto output_node_idx = (*it).Index();
+    output_nodes_idx.push_back(&output_node_idx);
+  }
 
-    for (int i = 0; i < static_cast<int>(new_dims.size()); i++) {
-      if (i < tensor_proto->dims().size()) {
-        new_tensor_proto.set_dims(i, new_dims[i]);
-      } else {
-        new_tensor_proto.add_dims(new_dims[i]);
-      }
-    }
-    graph.RemoveInitializedTensor(input_def->Name());
-    graph.AddInitializedTensor(new_tensor_proto);
+  // Remove output edges of the Unsqueeze node.
+  graph_utils::RemoveNodeOutputEdges(graph, node);
 
-    // Update shape of NodeArg
-    TensorShapeProto shape;
-    for (auto dim : new_dims) {
-      shape.add_dim()->set_dim_value(dim);
+  // Replace the input of the nodes following the Unsqueeze node.
+  const NodeArg* output_def = node.OutputDefs()[0];
+  for (NodeIndex* idx : output_nodes_idx) {
+    auto output_node = graph.GetNode(*idx);
+    if (!output_node) {
+      return Status(ONNXRUNTIME, INVALID_ARGUMENT);
     }
-    input_def->SetShape(shape);
-
-    // Replace the input of the nodes following unsqueeze node
-    const NodeArg* output_def = node.OutputDefs()[0];
-    for (auto it = node.OutputNodesBegin(); it != node.OutputNodesEnd(); ++it) {
-      auto output_node = graph.GetNode((*it).Index());
-      if (!output_node) {
-        return Status(ONNXRUNTIME, INVALID_ARGUMENT);
-      }
-
-      auto& input_defs = output_node->MutableInputDefs();
-      for (auto& def : input_defs) {
-        if (def == output_def) {
-          def = input_def;
-        }
+    auto& input_defs = output_node->MutableInputDefs();
+    for (auto& def : input_defs) {
+      if (def == output_def) {
+        def = input_def;
       }
     }
-
-    removed_nodes.push_back(node.Index());
-  }
-
-  for (auto i : removed_nodes) {
-    graph.RemoveNode(i);
   }
 
-  if (!removed_nodes.empty()) {
-    modified = true;
-  }
+  // Remove the Unsqueeze node.
+  graph.RemoveNode(node.Index());
+  removed = modified = true;
```

Comment:
>removed = modified  [](start = 2, length = 19)

will there be a case when removed and modified have different values? if not then why do we need 2?

Reply:
Yes, there will be when you do not remove a node, but you only modify it.
The reason we have a separate removed is so that we stop the rule application for that node if it was removed. If it was only altered, you can keep on applying rules for that node.
I had discussed with @skottmckay about substituting this with an enum that takes values unmodified/modified/removed, but we have not gotten to it yet.

Reply:
I agree enum sounds better... but if that is not in plan right away then for time being may be you can change deleted to more detailed naming like - stop_rule_execution or something similar... because there may be other conditions when the node is not deleted but you still want to stop further rule apply on that node...

This is just a suggestion.

---
In reply to: [267897090](https://github.com/Microsoft/onnxruntime/pull/670#discussion_r267897090) [](ancestors = 267897090)

Reply:
We can prioritize adding the enum if the current two booleans are confusing.
But I do not think we should call this stop_rule_execution here. What these booleans tell you is what the rewrite rule did to the graph. A naming like stop_rule_execution would be a "command" to the graph transformer manager, and the rewrite rule sits a level below it. It should give sufficient information for the transformer manager to decide what to do, but not tell it what to do.